src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -0,0 +1,2928 @@
+    if (builder.is_capability_required(SpvCapabilityInt8) && !target.has_feature(Target::VulkanInt8)) {
+        user_error << "Vulkan: Code requires 8-bit integer support (which is not enabled in the target features)! "
+                   << "Either enable the target feature, or adjust the algorithm to avoid using this data type!";
+    }
+
+    if (builder.is_capability_required(SpvCapabilityInt16) && !target.has_feature(Target::VulkanInt16)) {
+        user_error << "Vulkan: Code requires 16-bit integer support (which is not enabled in the target features)! "
+                   << "Either enable the target feature, or adjust the algorithm to avoid using this data type!";
+    }
+
+    if (builder.is_capability_required(SpvCapabilityInt64) && !target.has_feature(Target::VulkanInt64)) {
+        user_error << "Vulkan: Code requires 64-bit integer support (which is not enabled in the target features)! "
+                   << "Either enable the target feature, or adjust the algorithm to avoid using this data type!";
+    }
+
+    if (builder.is_capability_required(SpvCapabilityFloat16) && !target.has_feature(Target::VulkanFloat16)) {
+        user_error << "Vulkan: Code requires 16-bit floating-point support (which is not enabled in the target features)! "
+                   << "Either enable the target feature, or adjust the algorithm to avoid using this data type!";
+    }
+
+    if (builder.is_capability_required(SpvCapabilityFloat64) && !target.has_feature(Target::VulkanFloat64)) {
+        user_error << "Vulkan: Code requires 64-bit floating-point support (which is not enabled in the target features)! "
+                   << "Either enable the target feature, or adjust the algorithm to avoid using this data type!";
+    }
+
+    // Encode the SPIR-V into a compliant binary
+    SpvBinary spirv_binary;
+    builder.encode(spirv_binary);
+
+    size_t header_bytes = spirv_header.size() * sizeof(uint32_t);
+    size_t binary_bytes = spirv_binary.size() * sizeof(uint32_t);
+
+    debug(2) << "    encoding module ("
+             << "header_size: " << (uint32_t)(header_bytes) << ", "
+             << "binary_size: " << (uint32_t)(binary_bytes) << ")\n";
+
+    // Combine the header and binary into the module
+    module.reserve(header_bytes + binary_bytes);
+    module.insert(module.end(), (const char *)spirv_header.data(), (const char *)(spirv_header.data() + spirv_header.size()));
+    module.insert(module.end(), (const char *)spirv_binary.data(), (const char *)(spirv_binary.data() + spirv_binary.size()));
+}
+
+void CodeGen_Vulkan_Dev::SPIRV_Emitter::add_kernel(const Stmt &s,
+                                                   const std::string &kernel_name,
+                                                   const std::vector<DeviceArgument> &args) {
+    debug(2) << "Adding Vulkan kernel " << kernel_name << "\n";
+
+    // Add function definition
+    // TODO: can we use one of the function control annotations?
```

Comment:
Should this be an open issue, or removed?

Reply:
Added an open issue!